File: built-in plugins dsp/mb-restore-broadband.cpp
/**
 * MB Broadband Denoiser
 * Category : effect
 * Type     : gate
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Multi-band adaptive noise floor removal for broadband noise
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_RESTORE_BROADBAND_H
#define MB_RESTORE_BROADBAND_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbRestoreBroadband : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-restore-broadband";
    static constexpr const char* PLUGIN_NAME    = "MB Broadband Denoiser";
    static constexpr const char* PLUGIN_TYPE    = "gate";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = -50f;  // range [-80, -10]
    float bands = 8f;  // range [4, 16]
    float reduction = 15f;  // range [0, 40]
    float smoothing = 0.5f;  // range [0, 1]
    };

    MbRestoreBroadband() = default;
    ~MbRestoreBroadband() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, -80f, -10f);
        params.bands = std::clamp(params.bands, 4f, 16f);
        params.reduction = std::clamp(params.reduction, 0f, 40f);
        params.smoothing = std::clamp(params.smoothing, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Broadband Denoiser
        return input;
    }
};

#endif // MB_RESTORE_BROADBAND_H

File: built-in plugins dsp/mb-restore-broadband.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbRestoreBroadbandPlugin: PluginDefinition = { id: 'mb-restore-broadband', slug: 'mb-restore-broadband', name: 'MB Broadband Denoiser', category: 'effect', type: 'gate' as any, version: '1.0.0', description: 'Multi-band adaptive noise floor removal for broadband noise', author: 'Max Booster', parameters: [{ id: 'threshold', name: 'Threshold', type: 'float', defaultValue: -50, minValue: -80, maxValue: -10, automatable: true }, { id: 'bands', name: 'Bands', type: 'float', defaultValue: 8, minValue: 4, maxValue: 16, automatable: false }, { id: 'reduction', name: 'Reduction', type: 'float', defaultValue: 15, minValue: 0, maxValue: 40, automatable: true }, { id: 'smoothing', name: 'Smoothing', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { threshold: -50, bands: 8, reduction: 15, smoothing: 0.5 } };

export default MbRestoreBroadbandPlugin;
